pkg/engine/planner/logical/planner.go

```diff
@@ -0,0 +1,231 @@
+	case labels.MatchNotEqual:
+		return types.BinaryOpNotMatchStr
+	case labels.MatchRegexp:
+		return types.BinaryOpMatchRe
+	case labels.MatchNotRegexp:
+		return types.BinaryOpNotMatchRe
+	default:
+		panic("invalid match type")
+	}
+}
+
+func convertLineMatchType(op log.LineMatchType) types.BinaryOp {
+	switch op {
+	case log.LineMatchEqual:
+		return types.BinaryOpMatchStr
+	case log.LineMatchNotEqual:
+		return types.BinaryOpNotMatchStr
+	case log.LineMatchRegexp:
+		return types.BinaryOpMatchRe
+	case log.LineMatchNotRegexp:
+		return types.BinaryOpNotMatchRe
+	case log.LineMatchPattern:
+		return types.BinaryOpMatchPattern
+	case log.LineMatchNotPattern:
+		return types.BinaryOpNotMatchPattern
+	default:
+		panic("invalid match type")
+	}
+}
+
+func convertLineFilter(filter syntax.LineFilter) Value {
+	return &BinOp{
+		Left:  logColumnRef(),
+		Right: LiteralString(filter.Match),
+		Op:    convertLineMatchType(filter.Ty),
+	}
+}
+
+func convertLineFilterExpr(expr *syntax.LineFilterExpr) Value {
+	if expr.Left != nil {
+		op := types.BinaryOpAnd
+		if expr.IsOrChild {
+			op = types.BinaryOpOr
+		}
+		return &BinOp{
+			Left:  convertLineFilterExpr(expr.Left),
+			Right: convertLineFilter(expr.LineFilter),
+			Op:    op,
+		}
+	}
+	return convertLineFilter(expr.LineFilter)
+}
+
+func convertLabelFilter(expr log.LabelFilterer) (Value, error) {
+	switch e := expr.(type) {
+	case *log.BinaryLabelFilter:
+		op := types.BinaryOpOr
+		if e.And == true {
+			op = types.BinaryOpAnd
+		}
+		left, err := convertLabelFilter(e.Left)
+		if err != nil {
+			return nil, err
+		}
+		right, err := convertLabelFilter(e.Right)
+		if err != nil {
+			return nil, err
+		}
+		return &BinOp{Left: left, Right: right, Op: op}, nil
+	case *log.BytesLabelFilter:
+		return nil, fmt.Errorf("not implemented: %T", e)
+	case *log.NumericLabelFilter:
+		return nil, fmt.Errorf("not implemented: %T", e)
+	case *log.DurationLabelFilter:
+		return nil, fmt.Errorf("not implemented: %T", e)
+	case *log.NoopLabelFilter:
+		return nil, fmt.Errorf("not implemented: %T", e)
+	case *log.StringLabelFilter:
+		m := e.Matcher
+		return &BinOp{
+			Left:  &ColumnRef{Column: m.Name, Type: types.ColumnTypeAmbiguous},
+			Right: LiteralString(m.Value),
+			Op:    convertLabelMatchType(m.Type),
+		}, nil
+	case *log.LineFilterLabelFilter:
+		m := e.Matcher
+		return &BinOp{
+			Left:  &ColumnRef{Column: m.Name, Type: types.ColumnTypeAmbiguous},
+			Right: LiteralString(m.Value),
+			Op:    convertLabelMatchType(m.Type),
+		}, nil
+	}
+	return nil, fmt.Errorf("invalid label filter %T", expr)
+}
+
+func convertQueryRangeToPredicates(start, end int64) []*BinOp {
+	return []*BinOp{
+		{
+			Left:  timestampColumnRef(),
+			Right: LiteralUint64(uint64(start)),
+			Op:    types.BinaryOpGte,
+		},
+		{
+			Left:  timestampColumnRef(),
+			Right: LiteralUint64(uint64(end)),
+			Op:    types.BinaryOpLt,
+		},
+	}
+}
+
+func ConvertToLogicalPlan(params logql.Params) (*Plan, error) {
```

Comment:
nit: `logical.ConvertToLogicalPlan` might be a bit more verbose than is needed for clarity here. `logical.BuildPlan`? (Also, please add a doc comment 🙏)